File: frontend/src/types/path.ts
export interface SkillNode {
  id: string;
  name: string;
  name_ko: string;
  category: string;
  difficulty: number;
  market_demand: number;
  estimated_hours: number;
  description: string;
  source: string;
}

export interface PathNode extends SkillNode {
  status: "not_started" | "in_progress" | "completed" | "recommended_next";
  order: number;
  explanation: {
    why_needed: string;
    job_relevance: string;
    connection_to_next: string;
  };
}

export interface PathEdge {
  source: string;
  target: string;
  type: string;
  on_recommended_path: boolean;
  strength: number;
}

export interface PathMetadata {
  total_hours: number;
  total_nodes: number;
  skipped_count: number;
  skipped_skills: string[];
  path_type: "fast_track" | "fundamentals";
  job_title: string;
  concurrent_groups: string[][];
  goal_node_id?: string;
}

export interface GeneratedPath {
  nodes: PathNode[];
  edges: PathEdge[];
  metadata: PathMetadata;
}

export interface DualPath {
  fast_track: GeneratedPath;
  fundamentals: GeneratedPath;
  shared_graph: {
    nodes: SkillNode[];
    edges: PathEdge[];
  };
}

export type ProgressStatus = "not_started" | "in_progress" | "completed";

export type ProgressMap = Record<string, ProgressStatus>;

export interface PathResponse {
  id: string;
  job_input: string;
  user_state_input: string;
  paths: DualPath;
  progress: ProgressMap;
  created_at: string;
}

File: backend/app/schemas/path.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class PathGenerateRequest(BaseModel):
    job_input: str
    current_state: str


class PathNode(BaseModel):
    id: str
    name: str
    name_ko: str
    category: str
    difficulty: float
    market_demand: float
    estimated_hours: float
    description: str
    status: str
    order: int
    explanation: dict[str, Any]


class PathEdge(BaseModel):
    source: str
    target: str
    type: str
    on_recommended_path: bool
    strength: float


class PathMetadata(BaseModel):
    total_hours: float
    total_nodes: int
    skipped_count: int
    skipped_skills: list[str] = []  # Names of skills user already knows
    path_type: str
    job_title: str
    concurrent_groups: list[list[str]] = []  # Groups of skill IDs learnable in parallel


class GeneratedPath(BaseModel):
    nodes: list[PathNode]
    edges: list[PathEdge]
    metadata: PathMetadata


class DualPath(BaseModel):
    fast_track: GeneratedPath
    fundamentals: GeneratedPath
    shared_graph: dict[str, Any]


class PathResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_input: str
    user_state_input: str
    paths: DualPath
    created_at: datetime


class ProgressUpdate(BaseModel):
    status: str
